### ISSUE TITLE
[explicit-function-return-type] Don't report when using an explicit variable type with an arrow function

### ISSUE THREAD

**Issue by cheezenaan** [typescript-eslint/typescript-eslint#149]
**Repro**

```JSON
{
  "extends": ["plugin:@typescript-eslint/recommended"],
  "plugins": ["react"]
}
```

```TS
import * as React from 'react';

interface Props {
    message?: string;
}

export const Hello: React.SFC<Props> = ({ message = 'kiniro-mosaic' }) => (
    <div>{`Hello! ${message}`}</div>;
);

```

**Expected Result**

Neither errors nor warnings

**Actual Result**

```
7:40  warning  Missing return type on function  @typescript-eslint/explicit-function-return-type
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.1.0` |
| `@typescript-eslint/parser`        | `1.1.0` |
| `eslint-plugin-react`        | `7.12.4` |
| `react`                       | `16.7.0` |
| `react-dom`                       | `16.7.0` |
| `TypeScript`                       | `3.2.4` |
| `ESLint`                           | `5.12.1` |
| `node`                             | `8.11.3` |
| `npm`                              | `5.6.0` |

**Comment by codepunkt** [comment]
Is this the same problem?

```typescript
type Wat = () => string[]

// Missing return type on function.eslint(@typescript-eslint/explicit-function-return-type)
let fn: Wat = () => {
  return []
}
```

```typescript
type Wat = () => string[]

// no errors or warnings
let fn: Wat
fn = () => {
  return []
}
```

**Comment by gilbsgilbs** [comment]
@codepunkt Yes.

**Comment by codepunkt** [comment]
@gilbsgilbs can i help?

**Comment by gilbsgilbs** [comment]
@codepunkt I don't think so. I already made a [PR](https://github.com/typescript-eslint/typescript-eslint/pull/260) that add an option to ignore this rule when the function expression is typed. Just need to wait for merge and release.

**Comment by saranshkataria** [comment]
I see the latest changes have been published, but I still get an error when writing this:

const App: React.FunctionComponent<{}> = () => <div>Hello World</div>;

**Comment by gilbsgilbs** [comment]
@saranshkataria thanks. I think it was only released as alpha at the moment. Just to be sure:

- did you install the latest alpha release
- did you enable the `allowTypedFunctionExpressions` option on the rule? See https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-function-return-type.md#allowtypedfunctionexpressions

**Comment by bradzacher** [comment]
Yup, as @gilbsgilbs mentioned, this is only on the automatically released canary tag right now.
`yarn add @typescript-eslint/eslint-plugin@canary`

In future, please keep an eye on the releases section of the repo to see if PRs have released.
https://github.com/typescript-eslint/typescript-eslint/releases

**Comment by saranshkataria** [comment]
thanks for the info. I installed the alpha, and set 

`  rules: {
    allowTypedFunctionExpressions: true,
  }`

And still see it as a warning. Is that expected behavior?

**Comment by bradzacher** [comment]
maybe not? Hard to say without seeing your config/failing code.
Tests are green so either you've found an untested case, or your config is incorrect?

Could you please log it as a new issue?

**Comment by kkalavantavanich** [comment]
Is this the same issue? (`allowTypedFunctionExpressions` set to `true`)

```ts
type Bar = () => boolean;

const foo = (): Bar => {
  return () => true;
         ~~~~~~~~~~~  // Missing return type on function. eslint(@typescript-eslint/explicit-function-return-type)
};

foo();
```

**Comment by gilbsgilbs** [comment]
@kkalavantavanich I don't think so. But `allowExpressions: true` should work in this case.

**Comment by TidyIQ** [comment]
I'm still getting the error on canary release `"@typescript-eslint/eslint-plugin": "^1.7.1-alpha.17",`

Rules:
```
"rules": {
  "allowTypedFunctionExpressions": true,
  "react-hooks/rules-of-hooks": "error",
  "react-hooks/exhaustive-deps": "warn",
  "react/prop-types": "off"
},
```
I've also tried:
```
"rules": {
  "@typescript-eslint/allowTypedFunctionExpressions": true,
  "react-hooks/rules-of-hooks": "error",
  "react-hooks/exhaustive-deps": "warn",
  "react/prop-types": "off"
},
```

Code:
```
interface ToggleBoolean {
  (state: string, value: boolean): {
    type: string;
    payload: {
      state: string;
      value: boolean;
    };
  };
}
export const toggleBoolean: ToggleBoolean = (state, value) => {
  return {
    type: TOGGLE_BOOLEAN,
    payload: {
      state: state,
      value: value,
    },
  };
};
```

Error:
```
Missing return type on function.eslint(@typescript-eslint/explicit-function-return-type)
```

**Comment by TidyIQ** [comment]
Nevermind, I just saw [this issue](https://github.com/typescript-eslint/typescript-eslint/issues/384) that explains it. For anyone following along, the rule posted further up in the comments is incorrect. It should be:
```
  "rules": {
    "@typescript-eslint/explicit-function-return-type": ["error", {
      "allowTypedFunctionExpressions": true
    }],
  }
```

**Comment by ButterHub** [comment]
May I ask why we need to set a rule for this? What is the alternative, and why don't we have it ignore missing return types for react components?
